File: 1046 Last Stone Weight/Last Stone Weight.cpp
class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        priority_queue<int> maxHeap;
        for (int stone : stones) {
            maxHeap.push(stone);
        }
        
        while (maxHeap.size() > 1 && maxHeap.top() != 0) {
            int stone1 = maxHeap.top();
            maxHeap.pop();
            int stone2 = maxHeap.top();
            maxHeap.pop();
            maxHeap.push(stone1 - stone2);
        }
        
        return maxHeap.empty() ? 0 : maxHeap.top();
    }
};

File: 1046 Last Stone Weight/Last Stone Weight.js
/**
 * @param {number[]} stones
 * @return {number}
 */
var lastStoneWeight = function (stones) {
  const maxHeap = [...stones].sort((a, b) => b - a);

  while (maxHeap.length > 1 && maxHeap[0] !== 0) {
    const stone1 = maxHeap.shift();
    const stone2 = maxHeap.shift();
    maxHeap.push(stone1 - stone2);
    maxHeap.sort((a, b) => b - a);
  }

  return maxHeap[0];
};
